src/java.base/share/classes/java/io/Console.java

```diff
@@ -145,6 +146,124 @@ public Reader reader() {
         throw newUnsupportedOperationException();
     }
 
+    /**
+     * Writes a string representation of the specified object to this console's
+     * output stream, terminates the line and then flushes the console.
+     *
+     * <p> The effects of {@code con.println(obj)} are as if by calling
+     * {@snippet lang = java:
+     *     con.format("%s%n", obj)
+     * }
+     *
+     * @param  obj
+     *         An object whose string representation is to write.
+     *
+     * @return  This console
+     *
+     * @since 23
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.IMPLICIT_CLASSES)
+    public Console println(Object obj) {
+        throw newUnsupportedOperationException();
+    }
+
+    /**
+     * Writes a string representation of the specified object to this console's
+     * output stream and then flushes the console.
```

Comment:
I think the flushing behavior is worth a discussion. Certainly it should flush after printing a prompt and reading input in `readLine` and in `input`. Maybe it should be line buffered; that is, flush after printing every full line, such as after `println` but not necessarily after `print`. (Hm, what if the string contains an embedded newline?) Or maybe it should flush after every `print` or `format` call.

Since this is primarily intended for interactive use, flushing after every operation does seem reasonable. The main reason for buffering is to batch up output so that an expensive I/O operation can be amortized over more bytes. That doesn't really apply for interactive use. I can also see programmers performing some long computation and doing `print(".")` after each step, so as to illustrate progress. They won't see it if output is buffered. (Mostly just thinking out loud here.)

Anyway, we should make an explicit decision to flush after every operation, if that's what we decide to do.

Reply:
I agree that the flushing behaviour needs discussion. That said, I think we should address it separately: first, for new methods, then for the old ones.

I think that `Console.Basic` methods need to flush automatically, given their intended audience. As for the `Console` methods, both old and the new ones that this PR proposes, we need to think.

I'm not sure if it's true in all cases and, if it is, whether it's been always true, but `Console` behaves _as if_ it flushes automatically, meaning that a programmer does not need to call `Console.flush()`.

Reading this [guide][] and recollecting all the `System.out.println`s I've seen -- `System.out` is `PrintWriter` that also has `flush`, which I've never seen being called -- I'd imagine people's mental model is that those behave similarly: issue `printf` or `format` and you'll immediately see the result in the console without ever having to call `flush()`.

[guide]: https://docs.oracle.com/javase/tutorial/essential/io/cl.html

Reply:
Correction: `System.out` is `PrintStream`, not `PrintWriter`. Still, both of these classes are `Flushable` and that is what is important for our discussion.

While `Flushable` itself does not stipulate _when_ exactly one needs to call `flush`, `PrintStream` and `PrintWriter` do:

 * Optionally, a `PrintStream` can be created so as to flush automatically; this means that the `flush` method of the underlying output stream is automatically invoked after a byte array is written, one of the `println` methods is invoked, or a newline character or byte `('\n')` is written.

 * Unlike the `PrintStream` class, if automatic flushing is enabled it will be done only when one of the `println`, `printf`, or `format` methods is invoked, rather than whenever a newline character happens to be output. These methods use the platform's own notion of line separator rather than the newline character.

In practice, I cannot remember if I've ever seen a single call to flush on `System.out`'s `println`, `printf` or `print` for that matter. I think, by now, it's a common expectation that one shouldn't be required to `flush`.

Back to `Console`. `Console` is much less specific about flushing than `PrintStream` and `PrintWriter` are: aside from it also implementing `Flushable`, it says nothing about when one should `flush`. Yet, in my experiments and recollections, `System.console()` seems to be also flushing automatically.

I guess, the above paragraphs mean that if we suddenly stop providing "automatic flushing" even for some output operations, we will break a lot of code. This behaviour seems to have been out there for so long, that it's a part of the API now.

I don't know about `Console`, but unless we provide `flush` in `Console.Basic`, I think `Console.Basic` should not even mention flushing: all the `Console.Basic` methods must be self-sufficient operations.

Additionally, if we want `Console.Basic` methods to be simple 1-1 mapping to their namesake `Console` methods, rather than being combinations thereof, then the new methods in `Console` should specify that they automatically flush.

Here's an example to explain what I mean by mapping/combination. A simple 1-1 mapping call does this:

        public static void print(Object obj) {
            con().print(obj);
        }

whereas a combination could do this:

        public static void print(Object obj) {
            con().print(obj);
            con().flush;
        }

Reply:
In my cursory search through Maven Central, I wasn't able to find any artefact that calls to `System.out.flush()` aside from this in some _Kotlin_ build file in JUnit:

        if (result.exitValue != 0 && hideOutput.get()) {
            System.out.write(output.toByteArray())
            System.out.flush()
        }

However, in that same attempt, I found 16 artefacts that call to `Console.flush()`.

Reply:
OK. I'm gravitating toward a policy where, if the Console is attached to a terminal -- maybe defer to `isTerminal()`, though that seems somewhat ill-defined -- then flushing is done after every operation. Otherwise, the exact flushing behavior is unspecified. (Or perhaps in this case it's line-buffered.)